components/hab/src/command/pkg/binlink.rs

```diff
@@ -126,9 +154,96 @@ where
     Ok(())
 }
 
+fn is_dest_on_path<T: AsRef<Path>>(dest_dir: T) -> bool {
+    if let Some(val) = env::var_os("PATH") {
+        env::split_paths(&val).any(|p| p == dest_dir.as_ref())
+    } else {
+        false
+    }
+}
+
+struct Binlink {
+    dest: PathBuf,
+    src: PathBuf,
+}
+
+#[cfg(not(target_os = "windows"))]
+impl Binlink {
+    pub fn new<T: AsRef<Path>>(src: T, dest_dir: T) -> Self {
+        Self {
+            dest: dest_dir.as_ref().join(
+                src.as_ref()
+                    .file_name()
+                    .expect("Binlink source path does not resolve to a file"),
+            ),
+            src: src.as_ref().to_path_buf(),
+        }
+    }
+
+    pub fn from_file<T: AsRef<Path>>(dest: T) -> Result<Self> {
+        Ok(Binlink {
+            dest: dest.as_ref().to_path_buf(),
+            src: fs::read_link(&dest)?,
+        })
+    }
+
+    pub fn link(&self) -> Result<()> {
+        use hcore::os::filesystem;
+
+        filesystem::symlink(&self.src, &self.dest).map_err(Into::into)
+    }
+}
+
+#[cfg(target_os = "windows")]
+impl Binlink {
+    pub fn new<T: AsRef<Path>>(src: T, dest_dir: T) -> Self {
+        let mut path = dest_dir.as_ref().join(
+            src.as_ref()
+                .file_stem()
+                .expect("Binlink source path does not resolve to a file"),
+        );
+        path.set_extension("bat");
+
+        Binlink {
+            dest: path,
+            src: src.as_ref().to_path_buf(),
+        }
+    }
+
+    pub fn from_file<T: AsRef<Path>>(path: T) -> Result<Self> {
+        use toml::Value::Table;
+
+        let file = File::open(&path)?;
+        for line in BufReader::new(file).lines() {
+            let ln = line?;
+            if ln.to_uppercase().starts_with(BAT_COMMENT_MARKER) {
+                let (_, rest) = ln.split_at(BAT_COMMENT_MARKER.len());
+                if let Ok(Table(toml_exp)) = rest.parse() {
+                    if let Some(src) = toml_exp["source"].as_str() {
+                        return Ok(Binlink {
+                            dest: path.as_ref().to_path_buf(),
+                            src: PathBuf::from(src),
+                        });
+                    }
+                }
+            }
+        }
+        Err(Error::CannotParseBinlinkSource(path.as_ref().to_path_buf()))
+    }
+
+    pub fn link(&self) -> Result<()> {
+        let template = format!(
+            "@echo off\nREM source='{0}'\n\"{0}\" %*",
+            self.src.display()
+        );
+        Ok(fs::write(&self.dest, template)?)
```

Comment:
If you prefer `Ok(expression?)` over `expression.map_err(Into::into)`, that's fine with me, but either way, I think the windows and not(windows) versions should be consistent.

Reply:
I personally prefer (not super strongly) the `?` over `map_err`.